keps/sig-node/20190925-enable-seccomp-by-default.md

```diff
@@ -0,0 +1,335 @@
+
+ 
+### 4. Set `runtime/default-audit` as the default profile
+
+By default clusters will have `runtime/default-audit` applied to all containers (users can opt-out). This profile is similar to the existing `runtime/default`, with the difference being that syscalls used that are outside the allowed list will be logged instead of having their execution blocked. This improves the default auditing capabilities of high risk syscalls whilst not breaking existing workloads.
+
+An extra benefit is to provide an easier route for users to move to `runtime/default`. They can determine their workload readiness by going through violations on syslogs, instead of having to profile each application individually. 
+
+To support audit mode, some validation will be required ensuring the downstream dependencies (Kernel, Libseccomp and runc) also supports it. If audit mode is not supported, the default seccomp profile will be set back to `unconfined`, for backwards compatibility. 
+
+### User Stories
+
+#### Story 1
+As a user and administrator, I want to be able to assess what pods within my cluster would be affected by applying more restrictive seccomp profiles.
+
+#### Story 2
+As a user, I want to be able to assess whether my pods will be affected by a more restrictive seccomp profiles.
+
+#### Story 3
+As a user and administrator, I want to be able to create and maintain custom seccomp profiles without having to physically copy files across all my existing (and new) nodes.
+
+#### Story 4
+As a user and administrator, I want to be able Kubernetes to audit all potentially dangerous system calls from my containers by default.
+
+
+### Implementation Details
+
+### 1. Audit mode support (Details)
+
+Kubernetes will continue to be unaware of downstream support. If a user tries to use unsupported actions (i.e. `SCMP_ACT_LOG`) today, the lower level dependencies will return an error. As a result, the container won't be able to start. We are proposing no changes to this behaviour.
+
+
+### 2. Built-in profiles (Details)
+
+The internal built-in profiles will be implemented in golang, not allowing users to amend them. Similar to the implementation in the [containerd](https://github.com/containerd/containerd/blob/master/contrib/seccomp/seccomp_default.go) project.
+
+
+### 3. New Kind for custom profiles (Details)
+
+To represent the seccomp profiles a new `SeccompProfile` resource type will be created: 
+```
+// SeccompProfile defines custom seccomp profiles to be enforced at runtime.
+type SeccompProfile struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+
+	// +optional
+	Spec LinuxSeccomp
+}
+```
+
+Referencing the types defined by the [OCI Spec](https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L556):
+
+```
+// LinuxSeccomp represents syscall restrictions
+type LinuxSeccomp struct {
+	DefaultAction LinuxSeccompAction `json:"defaultAction"`
+	Architectures []Arch             `json:"architectures,omitempty"`
```

Comment:
I think in most cases this is irrelevant (i.e. cluster uses a single architecture). If the cluster does use different architectures, I think we can just have separate profiles.